zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkConnection.java

```diff
@@ -192,4 +230,53 @@ public String getServers() {
   public void addAuthInfo(String scheme, byte[] auth) {
     _zk.addAuthInfo(scheme, auth);
   }
+
+  private void lookupGetChildrenMethod() {
+    if (_getChildrenMethod != null) {
+      // Method is already cached.
+      return;
+    }
+    try {
+      if (GETCHILDREN_PAGINATION_DISABLED) {
+        lookupNonPaginatedGetChildren();
+      } else {
+        // Lookup the paginated getChildren API
+        _getChildrenMethod =
+            ZooKeeper.class.getMethod("getAllChildrenPaginated", String.class, boolean.class);
+      }
+    } catch (NoSuchMethodException e1) {
+      // Pagination API is not supported, fall back to non-paginated API
+      lookupNonPaginatedGetChildren();
+    }
+    LOG.info("Pagination config {}={}, method to be invoked: {}",
+        ZkSystemPropertyKeys.ZK_GETCHILDREN_PAGINATION_DISABLED, GETCHILDREN_PAGINATION_DISABLED,
+        _getChildrenMethod.getName());
+  }
+
+  private void lookupNonPaginatedGetChildren() {
+    try {
+      _getChildrenMethod = ZooKeeper.class.getMethod("getChildren", String.class, boolean.class);
+    } catch (NoSuchMethodException e2) {
+      // We should not expect this exception here.
+      LOG.error("getChildren is not supported in this zookeeper version!");
+    }
+  }
+
+  private void handleInvokedMethodException(Throwable cause)
+      throws KeeperException, InterruptedException {
+    if (cause instanceof KeeperException.UnimplementedException) {
+      LOG.warn("Paginated getChildren is unimplemented in ZK server! "
+          + "Falling back to non-paginated getChildren");
+      lookupNonPaginatedGetChildren();
```

Comment:
In this case, do you want to retry the operation?

Reply:
Right. That's the purpose of the next line `throw KeeperException.create(KeeperException.Code.CONNECTIONLOSS);`

When the `KeeperException.UnimplementedException` is thrown from the server, it is a connection loss, so just throw it to ZkClient and its retry logic will retry the `getChildren` operation appropriately.

Reply:
I think the retry should be done here with some fallback logic (which changes the cached method). Otherwise, the caller's retry will fail too, right? So why bother?

Reply:
The cached method is changed here, and later when zkClient retries calling `getChildren()`, it'll used the fallback API. 

We cannot retry calling zk server right now because the connection is loss. ZkClient has to wait for re-connected. So zkclient is the right place to retry the operation.

Reply:
This is hacky but I don't have any better way. Let's add a TODO here for the record. But the logic itself should be safe.

Reply:
@jiajunwang Here's the log (https://github.com/apache/helix/issues/1525#issuecomment-746900960) that proves the disconnection after `UnimplementedException` is returned from server.